core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -0,0 +1,255 @@
+        final Settings build = Settings.builder().put(this.settings).put(settings).build();
+        try {
+            List<RuntimeException> exceptions = new ArrayList<>();
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.prepareApply(build);
+                } catch (RuntimeException ex) {
+                    exceptions.add(ex);
+                    logger.debug("failed to prepareCommit settings for [{}]", ex, settingUpdater);
+                }
+            }
+            // here we are exhaustive and record all settings that failed.
+            ExceptionsHelper.rethrowAndSuppress(exceptions);
+        } finally {
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.rollback();
+                } catch (Exception e) {
+                    logger.error("failed to rollback settings for [{}]", e, settingUpdater);
+                }
+            }
+        }
+        return build;
+    }
+
+    /**
+     * Applies the given settings to all the settings consumers or to none of them. The settings
+     * will be merged with the node settings before they are applied while given settings override existing node
+     * settings.
+     * @param newSettings the settings to apply
+     * @return the unmerged applied settings
+    */
+    public synchronized Settings applySettings(Settings newSettings) {
+        if (lastSettingsApplied != null && newSettings.equals(lastSettingsApplied)) {
+            // nothing changed in the settings, ignore
+            return newSettings;
+        }
+        final Settings build = Settings.builder().put(this.settings).put(newSettings).build();
+        boolean success = false;
+        try {
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.prepareApply(build);
+                } catch (Exception ex) {
+                    logger.warn("failed to prepareCommit settings for [{}]", ex, settingUpdater);
+                    throw ex;
+                }
+            }
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                settingUpdater.apply();
+            }
+            success = true;
+        } catch (Exception ex) {
+            logger.warn("failed to apply settings", ex);
+            throw ex;
+        } finally {
+            if (success == false) {
+                for (SettingUpdater settingUpdater : settingUpdaters) {
+                    try {
+                        settingUpdater.rollback();
+                    } catch (Exception e) {
+                        logger.error("failed to refresh settings for [{}]", e, settingUpdater);
+                    }
+                }
+            }
+        }
+
+        try {
+            for (Map.Entry<String, String> entry : newSettings.getAsMap().entrySet()) {
```

Comment:
this feels weird here. Especially where this class will be the base class for index level settings. I wonder what is a good home for it - SettingsModule is better then here, but ideally we can find somewhere in the logging universe